core/src/main/java/com/linecorp/armeria/common/ClientCookieEncoder.java

```diff
@@ -0,0 +1,232 @@
+ */
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.linecorp.armeria.common.CookieUtil.add;
+import static com.linecorp.armeria.common.CookieUtil.addQuoted;
+import static com.linecorp.armeria.common.CookieUtil.stringBuilder;
+import static com.linecorp.armeria.common.CookieUtil.stripTrailingSeparator;
+import static com.linecorp.armeria.common.CookieUtil.stripTrailingSeparatorOrNull;
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import io.netty.util.internal.InternalThreadLocalMap;
+
+/**
+ * A <a href="http://tools.ietf.org/html/rfc6265">RFC6265</a> compliant cookie encoder for client side.
+ *
+ * <p>Note that multiple cookies are supposed to be sent at once in a single {@code "Cookie"} header.</p>
+ *
+ * <pre>{@code
+ * RequestHeaders headers =
+ *     RequestHeaders.of(HttpHeaderNames.COOKIE,
+ *                       ClientCookieEncoder.strict().encode("JSESSIONID", "1234"));
+ * }</pre>
+ *
+ * @see ClientCookieDecoder
+ */
+public final class ClientCookieEncoder extends CookieEncoder {
+
+    // Forked from netty-4.1.43
+
+    /**
+     * Strict encoder that validates that name and value chars are in the valid scope and (for methods that
+     * accept multiple cookies) sorts cookies into order of decreasing path length, as specified in RFC6265.
+     */
+    private static final ClientCookieEncoder STRICT = new ClientCookieEncoder(true);
+
+    /**
+     * Lax instance that doesn't validate name and value, and (for methods that accept multiple cookies) keeps
+     * cookies in the order in which they were given.
+     */
+    private static final ClientCookieEncoder LAX = new ClientCookieEncoder(false);
+
+    private static final Cookie[] EMPTY_COOKIES = new Cookie[0];
+
+    /**
+     * Returns the strict encoder that validates that name and value chars are in the valid scope and
+     * (for methods that accept multiple cookies) sorts cookies into order of decreasing path length,
+     * as specified in RFC6265.
+     */
+    public static ClientCookieEncoder strict() {
+        return STRICT;
+    }
+
+    /**
+     * Returns the lax encoder that doesn't validate name and value, and (for methods that accept multiple
+     * cookies) keeps cookies in the order in which they were given.
+     */
+    public static ClientCookieEncoder lax() {
+        return LAX;
+    }
+
+    private ClientCookieEncoder(boolean strict) {
+        super(strict);
+    }
+
+    /**
+     * Encodes the specified cookie into a {@code "Cookie"} header value.
+     *
+     * @param name the cookie name
+     * @param value the cookie value
+     * @return a Rfc6265 style Cookie header value
+     */
+    public String encode(String name, String value) {
```

Comment:
From what I can tell, `ClientCookieEncoder` and `ServerCookieEncoder` expose the same API. Does it make sense to instead only expose `CookieEncoder` interface with `clientStrict()`, `clientLax()`, etc factories and the public methods?

Reply:
They are expected to generate the header string(s) which adheres to different syntax, so I'm not sure if it's a good idea to expose a common interface between them.

What do you think about just adding a bunch of static methods to `Cookie` and then hide all encoders and decoders?

```java
    static String toCookieHeader(boolean strict, Cookie... cookies);
    static String toCookieHeader(boolean strict, Iterable<? extends Cookie> cookies);

    // Note that multiple `Cookie` headers are prohibited by RFC 6265
    // and thus we do not accept multiple headers.
    static Cookies fromCookieHeader(boolean strict, String cookieHeader);

    static List<String> toSetCookieHeaders(boolean strict, Cookie... cookies);
    static List<String> toSetCookieHeaders(boolean strict, Iterable<? extends Cookie> cookies);

    static Cookies fromSetCookieHeaders(boolean strict, String... setCookieHeaders);
    static Cookies fromSetCookieHeaders(boolean strict, Iterable<String> setCookieHeaders);

    // member methods
    String toCookieHeader(boolean strict);
    String toSetCookieHeader(boolean strict);
```

Reply:
Went ahead and:

- Added the static methods and member methods mentioned above.
  - Also added the versions without `boolean strict` which default to strict mode.
- Hid `{Client,Server}Cookie{Encoder,Decoder}` from the public API.